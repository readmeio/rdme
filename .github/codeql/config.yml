paths-ignore:
  - '__tests__/**'
