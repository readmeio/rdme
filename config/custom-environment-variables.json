{ "host": "README_HOST", "hub": "README_HUB" }
