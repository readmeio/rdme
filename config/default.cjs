module.exports = {
  // eslint-disable-next-line global-require
  cli: require('../package.json').name,
  host: 'https://dash.readme.com',
  hub: 'https://{project}.readme.io', // this is only used for the `open` command
};
