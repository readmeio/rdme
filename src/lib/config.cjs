const path = require('path');

/**
 * We have to do this otherwise `require('config')` loads from the cwd where the user is running
 * `rdme` which wont be what we want.
 *
 * This file also has a CJS extension because the `config` library uses `require()` calls and
 * because `rdme` has `type: module` in its `package.json` Node wants **everything** to use
 * `import`.
 */
const configDir = process.env.NODE_CONFIG_DIR;
process.env.NODE_CONFIG_DIR = path.join(__dirname, '../../config');

const config = require('config');

process.env.NODE_CONFIG_DIR = configDir;

module.exports = config;
