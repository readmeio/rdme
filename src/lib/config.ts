import pkg from '../../package.json';

const config = {
  cli: pkg.name,
  host: 'https://dash.readme.com',
  hub: 'https://{project}.readme.io', // this is only used for the `open` command
} as const;

export default config;
