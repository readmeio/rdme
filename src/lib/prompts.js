const semver = require('semver');
const { prompt } = require('enquirer');
const parse = require('parse-link-header');

exports.generatePrompts = (versionList, selectOnly = false) => [
  {
    type: 'select',
    name: 'option',
    message: 'Would you like to use an existing project version or create a new one?',
    skip() {
      return selectOnly;
    },
    choices: [
      { message: 'Use existing', value: 'update' },
      { message: 'Create a new version', value: 'create' },
    ],
  },
  {
    type: 'select',
    name: 'versionSelection',
    message: 'Select your desired version',
    skip() {
      return selectOnly ? false : this.enquirer.answers.option !== 'update';
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    skip() {
      return selectOnly ? true : this.enquirer.answers.option === 'update';
    },
    hint: '1.0.0',
  },
];

function specOptions(specList, parsedDocs, currPage, totalPages) {
  const specs = specList.map(s => {
    return {
      message: s.title,
      value: s._id, // eslint-disable-line no-underscore-dangle
    };
  });
  if (parsedDocs.prev.page) specs.push({ message: `< Prev (page ${currPage - 1} of ${totalPages})`, value: 'prev' });
  if (parsedDocs.next.page) {
    specs.push({ message: `Next (page ${currPage + 1} of ${totalPages}) >`, value: 'next' });
  }
  return specs;
}

const updateOasPrompt = (specList, parsedDocs, currPage, totalPages, getSpecs) => [
  {
    type: 'select',
    name: 'specId',
    message: 'Select your desired file to update',
    choices: specOptions(specList, parsedDocs, currPage, totalPages),
    async result(spec) {
      if (spec === 'prev') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.prev.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await newSpecs.json();
          const { specId } = await prompt(
            updateOasPrompt(newSpecList, newParsedDocs, currPage - 1, totalPages, getSpecs)
          );
          return specId;
        } catch (e) {
          return null;
        }
      }
      if (spec === 'next') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.next.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await newSpecs.json();
          const { specId } = await prompt(
            updateOasPrompt(newSpecList, newParsedDocs, currPage + 1, totalPages, getSpecs)
          );
          return specId;
        } catch (e) {
          return null;
        }
      }
      return spec;
    },
  },
];

exports.createOasPrompt = (specList, parsedDocs, totalPages, getSpecs) => [
  {
    type: 'select',
    name: 'option',
    message: 'Would you like to update an existing OAS file or create a new one?',
    choices: [
      { message: 'Update existing', value: 'update' },
      { message: 'Create a new spec', value: 'create' },
    ],
    async result(picked) {
      if (picked === 'update') {
        try {
          const { specId } = await prompt(updateOasPrompt(specList, parsedDocs, 1, totalPages, getSpecs));
          return specId;
        } catch (e) {
          return null;
        }
      }
      return picked;
    },
  },
];

exports.createVersionPrompt = (versionList, opts, isUpdate) => [
  {
    type: 'select',
    name: 'from',
    message: 'Which version would you like to fork from?',
    skip() {
      return opts.fork || isUpdate;
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    initial: opts.newVersion || false,
    skip() {
      return opts.newVersion || !isUpdate;
    },
    hint: '1.0.0',
    validate(val) {
      return semver.valid(semver.coerce(val)) ? true : this.styles.danger('Please specify a semantic version.');
    },
  },
  {
    type: 'confirm',
    name: 'is_stable',
    message: 'Would you like to make this version the main version for this project?',
    skip() {
      return opts.main || isUpdate?.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_beta',
    message: 'Should this version be in beta?',
    skip: () => opts.beta,
  },
  {
    type: 'confirm',
    name: 'is_hidden',
    message: 'Would you like to make this version public?',
    skip() {
      return opts.isPublic || opts.main || this.enquirer.answers.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_deprecated',
    message: 'Would you like to deprecate this version?',
    skip() {
      return opts.deprecated || opts.main || !isUpdate || this.enquirer.answers.is_stable;
    },
  },
];
