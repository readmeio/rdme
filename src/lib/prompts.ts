import type { Version } from '../cmds/versions';
import type { Response } from 'node-fetch';
import type { Choice, PromptObject } from 'prompts';

import parse from 'parse-link-header';
import semver from 'semver';

import promptTerminal from './promptWrapper';
import { handleRes } from './readmeAPIFetch';

interface Spec {
  _id: string;
  title: string;
}

type SpecList = Spec[];

interface ParsedDocs {
  next?: {
    page: number;
    url: string;
  };
  prev?: {
    page: number;
    url: string;
  };
}

function specOptions(specList: SpecList, parsedDocs: ParsedDocs, currPage: number, totalPages: number): Choice[] {
  const specs = specList.map(s => {
    return {
      description: `API Definition ID: ${s._id}`, // eslint-disable-line no-underscore-dangle
      title: s.title,
      value: s._id, // eslint-disable-line no-underscore-dangle
    };
  });
  if (parsedDocs?.prev?.page) {
    specs.push({
      description: 'Go to the previous page',
      title: `< Prev (page ${currPage - 1} of ${totalPages})`,
      value: 'prev',
    });
  }
  if (parsedDocs?.next?.page) {
    specs.push({
      description: 'Go to the next page',
      title: `Next (page ${currPage + 1} of ${totalPages}) >`,
      value: 'next',
    });
  }
  return specs;
}

const updateOasPrompt = (
  specList: SpecList,
  parsedDocs: ParsedDocs,
  currPage: number,
  totalPages: number,
  getSpecs: (url: string) => Promise<Response>,
): PromptObject[] => [
  {
    type: 'select',
    name: 'specId',
    message: 'Select your desired file to update',
    choices: specOptions(specList, parsedDocs, currPage, totalPages),
    async format(spec: string) {
      if (spec === 'prev') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.prev.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await handleRes(newSpecs);
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(
            updateOasPrompt(newSpecList, newParsedDocs, currPage - 1, totalPages, getSpecs),
          );
          return specId;
        } catch (e) {
          return null;
        }
      } else if (spec === 'next') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.next.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await handleRes(newSpecs);
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(
            updateOasPrompt(newSpecList, newParsedDocs, currPage + 1, totalPages, getSpecs),
          );
          return specId;
        } catch (e) {
          return null;
        }
      }

      return spec;
    },
  },
];

export function createOasPrompt(
  specList: SpecList,
  parsedDocs: ParsedDocs,
  totalPages: number,
  getSpecs: ((url: string) => Promise<Response>) | null,
): PromptObject[] {
  return [
    {
      type: 'select',
      name: 'option',
      message: 'Would you like to update an existing OAS file or create a new one?',
      choices: [
        { title: 'Update existing', value: 'update' },
        { title: 'Create a new spec', value: 'create' },
      ],
      async format(picked: 'update' | 'create') {
        if (picked === 'update') {
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(updateOasPrompt(specList, parsedDocs, 1, totalPages, getSpecs));
          return specId;
        }

        return picked;
      },
    },
  ];
}

/**
 * Series of prompts to construct a version object,
 * used in our `versions:create` and `versions:update` commands
 */
export function versionPrompt(
  /** list of versions, used for prompt about which version to fork */
  versionList: Version[],
  /** existing version if we're performing an update */
  isUpdate?: {
    is_stable: boolean;
  },
): PromptObject[] {
  return [
    {
      // only runs for versions:create command
      type: isUpdate ? null : 'select',
      name: 'from',
      message: 'Which version would you like to fork from?',
      choices: versionList.map(v => {
        return {
          title: v.version,
          value: v.version,
        };
      }),
    },
    {
      // only runs for versions:update command
      type: !isUpdate ? null : 'text',
      name: 'newVersion',
      message: 'What should the version be renamed to?',
      hint: '1.0.0',
      validate(val: string) {
        // allow empty string, in which case the version won't be renamed
        if (!val) return true;
        return semver.valid(semver.coerce(val)) ? true : 'Please specify a semantic version.';
      },
    },
    {
      // if the existing version being updated is already the main version,
      // we can't switch that so we skip this question
      type: isUpdate?.is_stable ? null : 'confirm',
      name: 'is_stable',
      message: 'Would you like to make this version the main version for this project?',
    },
    {
      type: 'confirm',
      name: 'is_beta',
      message: 'Should this version be in beta?',
    },
    {
      type: (prev, values) => {
        // if user wants this version to be the main version
        // it can't also be hidden.
        return values.is_stable ? null : 'confirm';
      },
      name: 'is_public',
      message: 'Would you like to make this version public?',
    },
    {
      type: (prev, values) => {
        // if user wants this version to be the main version
        // it can't also be deprecated.
        return values.is_stable ? null : 'confirm';
      },
      name: 'is_deprecated',
      message: 'Would you like to deprecate this version?',
    },
  ];
}
