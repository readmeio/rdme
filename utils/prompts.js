exports.generatePrompts = versionList => [
  {
    type: 'select',
    name: 'option',
    message:
      "We couldn't find a version in ReadMe matching the version in your OAS file. Would you like to use an existing version or create a new one?",
    choices: [
      { message: 'Use existing', value: 'update' },
      { message: 'Create a new version', value: 'create' },
    ],
  },
  {
    type: 'select',
    name: 'versionSelection',
    message: 'Select your desired version',
    skip() {
      return this.enquirer.answers.option !== 'update';
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
];

exports.createVersionPrompt = (versionList, opts, isUpdate) => [
  {
    type: 'select',
    name: 'from',
    message: 'Which version would you like to fork from?',
    skip() {
      return opts.fork || isUpdate;
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    skip() {
      return opts.newVersion || !isUpdate;
    },
    hint: '1.0.0',
  },
  {
    type: 'confirm',
    name: 'is_stable',
    message: 'Would you like to make this version the main version for this project?',
    skip() {
      return opts.main || (isUpdate && isUpdate.is_stable);
    },
  },
  {
    type: 'confirm',
    name: 'is_beta',
    message: 'Should this version be in beta?',
    skip: () => opts.beta,
  },
  {
    type: 'confirm',
    name: 'is_hidden',
    message: 'Would you like to make this version public?',
    skip() {
      return opts.isPublic || opts.main || this.enquirer.answers.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_deprecated',
    message: 'Would you like to deprecate this version?',
    skip() {
      return opts.deprecated || opts.main || !isUpdate || this.enquirer.answers.is_stable;
    },
  },
];
