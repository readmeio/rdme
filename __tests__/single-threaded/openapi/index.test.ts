/* eslint-disable no-console */
import fs from 'node:fs';

import chalk from 'chalk';
import { http } from 'msw';
import { setupServer } from 'msw/node';
import prompts from 'prompts';
import { describe, beforeAll, beforeEach, afterEach, it, expect, vi, afterAll } from 'vitest';

import OpenAPICommand from '../../../src/cmds/openapi/index.js';
import config from '../../../src/lib/config.js';
import { getAPIMockMSW } from '../../helpers/get-api-mock.js';
import { after, before } from '../../helpers/get-gha-setup.js';
import { after as afterGHAEnv, before as beforeGHAEnv } from '../../helpers/setup-gha-env.js';

const openapi = new OpenAPICommand();

let consoleInfoSpy;
let consoleWarnSpy;

const key = 'API_KEY';
const version = '1.0.0';
const exampleRefLocation = `${config.host}/project/example-project/1.0.1/refs/ex`;
const successfulMessageBase = (specPath, specType) => [
  '',
  `\t${chalk.green(exampleRefLocation)}`,
  '',
  `To update your ${specType} definition, run the following:`,
  '',
  `\t${chalk.green(`rdme openapi ${specPath} --key=<key> --id=1`)}`,
];
const successfulUpload = (specPath, specType = 'OpenAPI') =>
  [
    `You've successfully uploaded a new ${specType} file to your ReadMe project!`,
    ...successfulMessageBase(specPath, specType),
  ].join('\n');

const getCommandOutput = () => {
  return [consoleWarnSpy.mock.calls.join('\n\n'), consoleInfoSpy.mock.calls.join('\n\n')].filter(Boolean).join('\n\n');
};

const getRandomRegistryId = () => Math.random().toString(36).substring(2);

const server = setupServer(...[]);

describe('rdme openapi (single-threaded)', () => {
  let testWorkingDir: string;

  beforeAll(() => server.listen({ onUnhandledRequest: 'error' }));

  beforeEach(() => {
    consoleInfoSpy = vi.spyOn(console, 'info').mockImplementation(() => {});
    consoleWarnSpy = vi.spyOn(console, 'warn').mockImplementation(() => {});
    testWorkingDir = process.cwd();
  });

  afterEach(() => {
    consoleInfoSpy.mockRestore();
    consoleWarnSpy.mockRestore();

    server.resetHandlers();

    process.chdir(testWorkingDir);
  });

  afterAll(() => server.close());

  describe('upload', () => {
    it('should discover and upload an API definition if none is provided', async () => {
      expect.assertions(6);
      const registryUUID = getRandomRegistryId();

      server.use(
        ...[
          getAPIMockMSW(`/api/v1/version/${version}`, 200, { json: { version } }, key),
          http.post(`${config.host}/api/v1/api-registry`, async ({ request }) => {
            const body = await request.text();
            expect(body).toMatch('form-data; name="spec"');
            return Response.json({ registryUUID, spec: { openapi: '3.0.0' } }, { status: 201 });
          }),
          getAPIMockMSW('/api/v1/api-specification', 200, { json: [] }, key, { 'x-readme-version': version }),
          http.post(`${config.host}/api/v1/api-specification`, async ({ request }) => {
            const body = await request.json();
            expect(body).toStrictEqual({ registryUUID });
            expect(request.headers.get('authorization')).toBeBasicAuthApiKey(key);
            return Response.json({ _id: 1 }, { status: 201, headers: { location: exampleRefLocation } });
          }),
        ],
      );

      const spec = 'petstore.json';

      await expect(
        openapi.run({
          key,
          version,
          workingDirectory: './__tests__/__fixtures__/relative-ref-oas',
        }),
      ).resolves.toBe(successfulUpload(spec));

      expect(console.info).toHaveBeenCalledTimes(1);

      const output = getCommandOutput();
      return expect(output).toBe(chalk.yellow(`ℹ️  We found ${spec} and are attempting to upload it.`));
    });

    it('should use specified working directory and upload the expected content', async () => {
      expect.assertions(6);
      let requestBody;
      const registryUUID = getRandomRegistryId();

      server.use(
        ...[
          getAPIMockMSW(`/api/v1/version/${version}`, 200, { json: { version } }, key),
          http.post(`${config.host}/api/v1/api-registry`, async ({ request }) => {
            const body = await request.text();
            requestBody = body.substring(body.indexOf('{'), body.lastIndexOf('}') + 1);
            requestBody = JSON.parse(requestBody);
            expect(body).toMatch('form-data; name="spec"');
            return Response.json({ registryUUID, spec: { openapi: '3.0.0' } }, { status: 201 });
          }),
          getAPIMockMSW('/api/v1/api-specification', 200, { json: [] }, key, { 'x-readme-version': version }),
          http.post(`${config.host}/api/v1/api-specification`, async ({ request }) => {
            const body = await request.json();
            expect(body).toStrictEqual({ registryUUID });
            expect(request.headers.get('authorization')).toBeBasicAuthApiKey(key);
            return Response.json({ _id: 1 }, { status: 201, headers: { location: exampleRefLocation } });
          }),
        ],
      );

      const spec = 'petstore.json';

      await expect(
        openapi.run({
          spec,
          key,
          version,
          workingDirectory: './__tests__/__fixtures__/relative-ref-oas',
        }),
      ).resolves.toBe(successfulUpload(spec));

      expect(console.info).toHaveBeenCalledTimes(0);

      return expect(requestBody).toMatchSnapshot();
    });

    it('should return spec create info for dry run', async () => {
      expect.assertions(3);
      const registryUUID = getRandomRegistryId();

      server.use(
        ...[
          getAPIMockMSW(`/api/v1/version/${version}`, 200, { json: { version } }, key),
          http.post(`${config.host}/api/v1/api-registry`, async ({ request }) => {
            const body = await request.text();
            expect(body).toMatch('form-data; name="spec"');
            return Response.json({ registryUUID, spec: { openapi: '3.0.0' } }, { status: 201 });
          }),
          getAPIMockMSW('/api/v1/api-specification', 200, { json: [] }, key, { 'x-readme-version': version }),
        ],
      );

      await expect(
        openapi.run({
          key,
          version,
          dryRun: true,
          workingDirectory: './__tests__/__fixtures__/relative-ref-oas',
        }),
      ).resolves.toMatch(
        '🎭 dry run! The API Definition located at petstore.json will be created for this project version: 1.0.0',
      );

      const output = getCommandOutput();
      return expect(output).toMatch(
        chalk.yellow('🎭 dry run option detected! No API definitions will be created or updated in ReadMe.'),
      );
    });
  });

  describe('error handling', () => {
    it('should error if no file was provided or able to be discovered', () => {
      return expect(openapi.run({ key, version, workingDirectory: 'bin' })).rejects.toStrictEqual(
        new Error(
          "We couldn't find an OpenAPI or Swagger definition.\n\nPlease specify the path to your definition with `rdme openapi ./path/to/api/definition`.",
        ),
      );
    });
  });

  describe('GHA onboarding E2E tests', () => {
    let yamlOutput;

    beforeEach(() => {
      before((fileName, data) => {
        yamlOutput = data;
      });
    });

    afterEach(() => {
      after();
    });

    it('should create GHA workflow (including workingDirectory)', async () => {
      expect.assertions(7);
      const yamlFileName = 'openapi-file-workingdirectory';
      prompts.inject([true, 'openapi-branch-workingdirectory', yamlFileName]);
      const registryUUID = getRandomRegistryId();

      server.use(
        ...[
          getAPIMockMSW(`/api/v1/version/${version}`, 200, { json: { version } }, key),
          http.post(`${config.host}/api/v1/api-registry`, async ({ request }) => {
            const body = await request.text();
            expect(body).toMatch('form-data; name="spec"');
            return Response.json({ registryUUID, spec: { openapi: '3.0.0' } }, { status: 201 });
          }),
          getAPIMockMSW('/api/v1/api-specification', 200, { json: [] }, key, { 'x-readme-version': version }),
          http.post(`${config.host}/api/v1/api-specification`, async ({ request }) => {
            expect(request.headers.get('authorization')).toBeBasicAuthApiKey(key);
            const body = await request.json();
            expect(body).toStrictEqual({ registryUUID });
            return Response.json({ _id: 1 }, { status: 201, headers: { location: exampleRefLocation } });
          }),
        ],
      );

      const spec = 'petstore.json';

      await expect(
        openapi.run({
          spec,
          key,
          version,
          workingDirectory: './__tests__/__fixtures__/relative-ref-oas',
        }),
      ).resolves.toMatchSnapshot();

      expect(yamlOutput).toMatchSnapshot();
      expect(fs.writeFileSync).toHaveBeenCalledTimes(2);
      return expect(fs.writeFileSync).toHaveBeenNthCalledWith(
        2,
        `.github/workflows/${yamlFileName}.yml`,
        expect.any(String),
      );
    });
  });

  describe('command execution in GitHub Actions runner', () => {
    beforeEach(() => {
      beforeGHAEnv();
    });

    afterEach(afterGHAEnv);

    it('should contain request header with correct URL with working directory', async () => {
      expect.assertions(8);
      const registryUUID = getRandomRegistryId();
      server.use(
        ...[
          getAPIMockMSW(`/api/v1/version/${version}`, 200, { json: { version } }, key),
          http.post(`${config.host}/api/v1/api-registry`, async ({ request }) => {
            const body = await request.text();
            expect(body).toMatch('form-data; name="spec"');
            return Response.json({ registryUUID, spec: { openapi: '3.0.0' } }, { status: 201 });
          }),
          getAPIMockMSW('/api/v1/api-specification', 200, { json: [] }, key, { 'x-readme-version': version }),
          http.post(`${config.host}/api/v1/api-specification`, async ({ request }) => {
            expect(request.headers.get('authorization')).toBeBasicAuthApiKey(key);
            expect(request.headers.get('x-rdme-ci')).toBe('GitHub Actions (test)');
            expect(request.headers.get('x-readme-source')).toBe('cli-gh');
            expect(request.headers.get('x-readme-source-url')).toBe(
              'https://github.com/octocat/Hello-World/blob/ffac537e6cbbf934b08745a378932722df287a53/__tests__/__fixtures__/relative-ref-oas/petstore.json',
            );
            expect(request.headers.get('x-readme-version')).toBe(version);
            const body = await request.json();
            expect(body).toStrictEqual({ registryUUID });
            return Response.json({ _id: 1 }, { status: 201, headers: { location: exampleRefLocation } });
          }),
        ],
      );

      const spec = 'petstore.json';

      await expect(
        openapi.run({
          spec,
          key,
          version,
          workingDirectory: './__tests__/__fixtures__/relative-ref-oas',
        }),
      ).resolves.toBe(successfulUpload(spec));

      return after();
    });
  });
});
