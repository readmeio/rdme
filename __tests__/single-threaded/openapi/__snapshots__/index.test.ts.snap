// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rdme openapi (single-threaded) > GHA onboarding E2E tests > should create GHA workflow (including workingDirectory) 1`] = `
"
Your GitHub Actions workflow file has been created! ✨

Almost done! Just a couple more steps:
1. Push your newly created file (.github/workflows/openapi-file-workingdirectory.yml) to GitHub 🚀
2. Create a GitHub secret called README_API_KEY and populate the value with your ReadMe API key (••••••••••••I_KEY) 🔑

🔐 Check out GitHub's docs for more info on creating encrypted secrets (https://docs.github.com/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-a-repository)

🦉 If you have any more questions, feel free to drop us a line! support@readme.io
"
`;

exports[`rdme openapi (single-threaded) > GHA onboarding E2E tests > should create GHA workflow (including workingDirectory) 2`] = `
"# This GitHub Actions workflow was auto-generated by the \`rdme\` cli on 2022-01-01T00:00:00.000Z
# You can view our full documentation here: https://docs.readme.com/docs/rdme
name: ReadMe GitHub Action 🦉

on:
  push:
    branches:
      # This workflow will run every time you push code to the following branch: \`openapi-branch-workingdirectory\`
      # Check out GitHub's docs for more info on configuring this:
      # https://docs.github.com/actions/using-workflows/events-that-trigger-workflows
      - openapi-branch-workingdirectory

jobs:
  rdme-openapi:
    runs-on: ubuntu-latest
    steps:
      - name: Check out repo 📚
        uses: actions/checkout@v3

      - name: Run \`openapi\` command 🚀
        uses: readmeio/rdme@v7
        with:
          rdme: openapi petstore.json --key=\${{ secrets.README_API_KEY }} --id=1 --workingDirectory=./__tests__/__fixtures__/relative-ref-oas
"
`;

exports[`rdme openapi (single-threaded) > upload > should use specified working directory and upload the expected content 1`] = `
{
  "components": {
    "securitySchemes": {
      "api_key": {
        "in": "header",
        "name": "api_key",
        "type": "apiKey",
      },
      "petstore_auth": {
        "flows": {
          "implicit": {
            "authorizationUrl": "http://petstore.swagger.io/oauth/dialog",
            "scopes": {
              "read:pets": "read your pets",
              "write:pets": "modify pets in your account",
            },
          },
        },
        "type": "oauth2",
      },
    },
  },
  "info": {
    "title": "Example petstore to demo our handling of relative external $ref pointers",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/pet": {
      "post": {
        "description": "",
        "operationId": "addPet",
        "requestBody": {
          "$ref": "#/paths/~1pet/put/requestBody",
        },
        "responses": {
          "405": {
            "description": "Invalid input",
          },
        },
        "security": [
          {
            "petstore_auth": [
              "write:pets",
              "read:pets",
            ],
          },
        ],
        "summary": "Add a new pet to the store",
        "tags": [
          "pet",
        ],
      },
      "put": {
        "description": "",
        "operationId": "updatePet",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/paths/~1pet~1%7BpetId%7D/get/responses/200/content/application~1xml/schema",
              },
            },
            "application/xml": {
              "schema": {
                "$ref": "#/paths/~1pet~1%7BpetId%7D/get/responses/200/content/application~1xml/schema",
              },
            },
          },
          "description": "Pet object that needs to be added to the store",
          "required": true,
        },
        "responses": {
          "400": {
            "description": "Invalid ID supplied",
          },
          "404": {
            "description": "Pet not found",
          },
          "405": {
            "description": "Validation exception",
          },
        },
        "security": [
          {
            "petstore_auth": [
              "write:pets",
              "read:pets",
            ],
          },
        ],
        "summary": "Update an existing pet",
        "tags": [
          "pet",
        ],
      },
    },
    "/pet/{petId}": {
      "get": {
        "description": "Returns a single pet",
        "operationId": "getPetById",
        "parameters": [
          {
            "description": "ID of pet to return",
            "in": "path",
            "name": "petId",
            "required": true,
            "schema": {
              "format": "int64",
              "type": "integer",
            },
          },
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/paths/~1pet~1%7BpetId%7D/get/responses/200/content/application~1xml/schema",
                },
              },
              "application/xml": {
                "schema": {
                  "properties": {
                    "category": {
                      "properties": {
                        "id": {
                          "format": "int64",
                          "type": "integer",
                        },
                        "name": {
                          "type": "string",
                        },
                      },
                      "type": "object",
                      "xml": {
                        "name": "Category",
                      },
                    },
                    "id": {
                      "default": 40,
                      "example": 25,
                      "format": "int64",
                      "type": "integer",
                    },
                    "name": {
                      "example": "doggie",
                      "type": "string",
                    },
                    "photoUrls": {
                      "items": {
                        "example": "https://example.com/photo.png",
                        "type": "string",
                      },
                      "type": "array",
                      "xml": {
                        "name": "photoUrl",
                        "wrapped": true,
                      },
                    },
                    "status": {
                      "description": "pet status in the store",
                      "enum": [
                        "available",
                        "pending",
                        "sold",
                      ],
                      "type": "string",
                    },
                    "tags": {
                      "items": {
                        "properties": {
                          "id": {
                            "format": "int64",
                            "type": "integer",
                          },
                          "name": {
                            "type": "string",
                          },
                        },
                        "type": "object",
                        "xml": {
                          "name": "Tag",
                        },
                      },
                      "type": "array",
                      "xml": {
                        "name": "tag",
                        "wrapped": true,
                      },
                    },
                  },
                  "required": [
                    "name",
                    "photoUrls",
                  ],
                  "type": "object",
                  "xml": {
                    "name": "Pet",
                  },
                },
              },
            },
            "description": "successful operation",
          },
          "400": {
            "description": "Invalid ID supplied",
          },
          "404": {
            "description": "Pet not found",
          },
          "default": {
            "description": "successful response",
          },
        },
        "security": [
          {
            "api_key": [],
          },
        ],
        "summary": "Find pet by ID",
        "tags": [
          "pet",
        ],
      },
    },
    "/user": {
      "post": {
        "description": "This can only be done by the logged in user.",
        "operationId": "createUser",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "properties": {
                  "email": {
                    "type": "string",
                  },
                  "firstName": {
                    "type": "string",
                  },
                  "id": {
                    "format": "int64",
                    "type": "integer",
                  },
                  "lastName": {
                    "type": "string",
                  },
                  "password": {
                    "type": "string",
                  },
                  "phone": {
                    "type": "string",
                  },
                  "userStatus": {
                    "description": "User Status",
                    "format": "int32",
                    "type": "integer",
                  },
                  "username": {
                    "type": "string",
                  },
                },
                "type": "object",
                "xml": {
                  "name": "User",
                },
              },
            },
          },
          "description": "Created user object",
          "required": true,
        },
        "responses": {
          "default": {
            "description": "successful operation",
          },
        },
        "summary": "Create user",
        "tags": [
          "user",
        ],
      },
    },
  },
  "servers": [
    {
      "url": "http://petstore.swagger.io/v2",
    },
  ],
}
`;
